sdks/python/apache_beam/typehints/schemas.py

```diff
@@ -0,0 +1,217 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+""" Support for mapping python types to proto Schemas and back again.
+
+Python              Schema
+np.int8     <-----> BYTE
+np.int16    <-----> INT16
+np.int32    <-----> INT32
+np.int64    <-----> INT64
+int         ---/
+np.float32  <-----> FLOAT
+np.float64  <-----> DOUBLE
+float       ---/
+bool        <-----> BOOLEAN
+
+The mappings for STRING and BYTES are different between python 2 and python 3,
+because of the changes to str:
+py3:
+str/unicode <-----> STRING
+bytes       <-----> BYTES
+ByteString  ---/
+
+py2:
+unicode     <-----> STRING
+str/bytes   ---/
+ByteString  <-----> BYTES
+"""
+
+from __future__ import absolute_import
+
+import sys
+from typing import ByteString
+from typing import Mapping
+from typing import NamedTuple
+from typing import Optional
+from typing import Sequence
+from uuid import uuid4
+
+import numpy as np
+from past.builtins import unicode
+
+from apache_beam.portability.api import schema_pb2
+from apache_beam.typehints.native_type_compatibility import _get_args
+from apache_beam.typehints.native_type_compatibility import _match_is_exactly_mapping
+from apache_beam.typehints.native_type_compatibility import _match_is_named_tuple
+from apache_beam.typehints.native_type_compatibility import _match_is_optional
+from apache_beam.typehints.native_type_compatibility import _safe_issubclass
+from apache_beam.typehints.native_type_compatibility import extract_optional_type
+
+
+# Registry of typings for a schema by UUID
+class SchemaTypeRegistry(object):
+  def __init__(self):
+    self.by_id = {}
+    self.by_typing = {}
+
+  def add(self, typing, schema):
+    self.by_id[schema.id] = (typing, schema)
+
+  def get_typing_by_id(self, unique_id):
+    result = self.by_id.get(unique_id, None)
+    return result[0] if result is not None else None
+
+  def get_schema_by_id(self, unique_id):
+    result = self.by_id.get(unique_id, None)
+    return result[1] if result is not None else None
+
+
+SCHEMA_REGISTRY = SchemaTypeRegistry()
+
+
+# Bi-directional mappings
+_PRIMITIVES = (
+    (np.int8, schema_pb2.AtomicType.BYTE),
+    (np.int16, schema_pb2.AtomicType.INT16),
+    (np.int32, schema_pb2.AtomicType.INT32),
+    (np.int64, schema_pb2.AtomicType.INT64),
+    (np.float32, schema_pb2.AtomicType.FLOAT),
+    (np.float64, schema_pb2.AtomicType.DOUBLE),
+    (unicode, schema_pb2.AtomicType.STRING),
+    (bool, schema_pb2.AtomicType.BOOLEAN),
+    (bytes if sys.version_info.major >= 3 else ByteString,
+     schema_pb2.AtomicType.BYTES),
+)
+
+PRIMITIVE_TO_ATOMIC_TYPE = dict((typ, atomic) for typ, atomic in _PRIMITIVES)
+ATOMIC_TYPE_TO_PRIMITIVE = dict((atomic, typ) for typ, atomic in _PRIMITIVES)
+
+# One-way mappings
+PRIMITIVE_TO_ATOMIC_TYPE.update({
+    # In python 3, this is a no-op because str == unicode,
+    # but in python 2 it overrides the bytes -> BYTES mapping.
+    str: schema_pb2.AtomicType.STRING,
```

Comment:
So the runtime types don't necessarily line up exactly with the typing representation of the schema. For example even though a schema may have an attribute with `np.int*` type, we still actually produce and consume `int` instances, and never use `np.int*` instances at runtime.

In this case, the typing might say `str`, but RowCoder uses StrUtf8Coder to produce/consume instances of `past.builtins.unicode` at runtime.

I agree this could be a little confusing for users. We discussed it [on the ML](https://lists.apache.org/thread.html/a26ebf52a6bfbd5035a77b41f870675b490311f0ba46212f9233249c@%3Cdev.beam.apache.org%3E) and @robertwb suggested this approach:

> In both Python 2 and Python 3 one would use str for STRING, it would decode to
past.builtins.unicode. This seems to capture the intent better than
mapping str to BYTES in Python 2 only.)


There are tests over in row_coder_test.py, and in standard_coders.yaml/standard_coders_test.py

Reply:
Thinking about this some more, what about just rejecting str in Python 2 (forcing the user to say unicode). We can loosen things if this becomes too cumbersome in the future (but going the other way is backwards incompatible).

Reply:
Do we really need to support python2? If this is going to be a burden in general, I would rather not add support for it.

Reply:
I think it'll be harder to consistently exclude support for Python 2 for all schema use.

Reply:
I think that makes sense. It doesn't seem too onerous to ask people to use unicode in python 2. And that's a good point that it's a backwards compatible change if we find out otherwise. I pushed a commit that does this: https://github.com/apache/beam/pull/9188/commits/ecaf73cd8eb82739c21f0a3551efa0d03cc842b6